### PR TITLE
Redesign `on` state in menu and dropdowns

### PR DESCRIPTION
**Pull request by Mati365**
### Suggested merge commit message ([convention](https://ckeditor.com/docs/ckeditor5/latest/framework/contributing/git-commit-message-convention.html))

Fix (ui): Improve accessibility of menu bar toggle items by marking them as `menucheckbox` aria roles. 
Feature (ui): Redesign of menu items styling.
 
Closes https://github.com/ckeditor/ckeditor5/issues/16572.

## Additional information

Commercial PR: https://github.com/cksource/ckeditor5-commercial/pull/6339

### Before

obraz
obraz

### After

obraz
obraz

### CODE REVIEW

**Comment by Mati365** [comment]
@scofalik Thx! Will be fixed.

**Comment by scofalik** [comment]
One more thing @Comandeer @oleq.

When I click the top-level menu item (like "Insert") it gets focus and the focus will follow the mouse hover. I wonder about this behavior. On one hand, it seems correct to show the focus when I clicked the top-level menu item. On the other hand, it is weird that the focus follows the mouse. But it would be also weird if it didn't. So I wonder if we have to show the focus on click and if we can hide it when hovering other items. It is also inconsistent with hovering items inside such menu. E.g. if I hover&click "Insert" it shows the focus but then I hover down and the focus is not shown anymore. But if I hover left&right, the focus is kept. See video.


https://github.com/ckeditor/ckeditor5/assets/1502228/10b3ef9a-a0a6-48c3-acad-449786ef7ae0

File: packages/ckeditor5-ui/src/menubar/utils.ts [hunk 1]
```diff
@@ -163,6 +163,32 @@ export const MenuBarBehaviors = {
...
+
+		menuBarView.listenTo( menuBarView.element!, 'keydown', () => {
+			isKeyPressed = true;
+		}, { useCapture: true } );
+
+		menuBarView.listenTo( menuBarView.element!, 'keyup', () => {
+			isKeyPressed = false;
+		}, { useCapture: true } );
+
+		menuBarView.listenTo( menuBarView.element!, 'focus', () => {
```
Comment (scofalik) [inline packages/ckeditor5-ui/src/menubar/utils.ts:187]: I just accidentally lost a long comment, so I am super happy to write it again :|.

/cc @oleq

So, the idea here is that: if focus has changed and keyboard key was pressed, then it means that this keyboard interaction caused the focus change and we need to start displaying outline.

This is an interesting, simple and smart idea. But I am not sure if I like it in the end.

First, it does not work for enter key (see my other comment).

Second, it grabs false positives. E.g. if I have shift key pressed and start hovering over menus, the outline gets displayed.

If we don't find a better and still reasonably simple solution then I am fine with that. But maybe there's some room for improvement. @oleq if you think that current solution is good enough we can leave it as is.

So, we can go one of two directions:

1.  Try to improve what we have. Try to solve enter key bug and maybe introduce list of keys that trigger this behavior (will fix false positives but may introduce false negatives if we introduce a behavior for a new key and forget to update the list).
2.  Try to implement `hadKeyboardFocusInteraction` handling precisely when it needs to be handled. This means going through all behaviors functions, possibly propagating something "up" to the menu bar component. I also think there's no behavior function for "open on enter" behavior and that's wired to the `execute` event, which again may be difficult to handle. I am afraid this will take us some time and ROI may be too small.

Again, @oleq please let me know what you think.

File: packages/ckeditor5-ui/src/dropdown/utils.ts [hunk 2]
```diff
@@ -541,6 +543,24 @@ function bindViewCollectionItemsToDefinitions(
 	definitions: Collection<ListDropdownItemDefinition>,
 	locale: Locale
 ) {
+	// List item checkboxes have a reserved space for the check icon, so we need to know if there are any checkboxes in the list
+	// to adjust the layout accordingly. It'd look weird if the items on the list were not aligned horizontally.
+	listItems.on( 'change', () => {
```
[the remark this answers is not in the document]
Comment (scofalik) [reply packages/ckeditor5-ui/src/dropdown/utils.ts:548]: Also, I wonder if we could instead listen to `add` and `remove` (or was it `delete`?) events on these collections and maybe keep internal counter of "toggleable" items. Or even simpler, just check if there was ever a toggleable button added, maybe such simplification will be good enough (to be honest - I am quite hesitant whether we need to have it dynamic; I think that it would be enough to set it "on initialization").

File: packages/ckeditor5-ui/src/menubar/menubarmenulistview.ts [hunk 3]
```diff
@@ -27,5 +32,64 @@ export default class MenuBarMenuListView extends ListView {
...
 
 		this.role = 'menu';
+		this.items.on( 'change', this._preallocateCheckSpacerInItemsIfNeeded.bind( this ) );
+	}
+
+	/**
+	 * This method adds empty space if there is any toggleable item in the list.
+	 * It makes the list properly aligned.
+	 */
+	private _preallocateCheckSpacerInItemsIfNeeded() {
```
Comment (scofalik) [inline packages/ckeditor5-ui/src/menubar/menubarmenulistview.ts:42]: I understand we have some code duplication, because similar code already exists but it's in "dropdown" space, not "menubar" space.

This code here for some reason looks overcomplicated. I somehow like the other solution better and I wonder what's the difference here, as the structures are more-or-less similiar. In the "dropdowns code" we also iterate through the items list, but we only need to check this:

```javascript
item instanceof ListItemView && item.children.first instanceof ListItemButtonView
```

to decide if that's a button or not. Why it is much more complicated here (you need two helper functions)? I also like in the other function that you first compile the list of buttons, and then check if any is toggleable, and then check the property for all buttons. Less places where you use "weird" functions like `pickListButtonMenuViewIfPresent()`.

I also wonder if we have a performance problem here due to this being called on every `change` callback. Are all items added together, or are they added one by one?

**Review by scofalik — CHANGES_REQUESTED** [review]
Other than what I commented, please try to use properties and methods name as short as possible. Maybe you'll find some room for improvement. I also dislike `maybeXXX` approach. I get your point, but that's not really necessary if we have typed code and if the method is very short or you immediately early return like this:

```js
const listView = pickListView();

if ( !listView ) {
    return;
}
```

**Comment by scofalik** [comment]
@oleq I haven't checked CSS files -- first, whether class names are correct regarding our naming convention; two, whether the CSS declarations are in correct files, in correct places, and, in general, make sense. I guess there were some CSS related bugs, so we can keep this review for later, but that's one more thing I'd like to ask you to check.

File: packages/ckeditor5-ui/src/button/buttonview.ts [hunk 4]
```diff
@@ -277,6 +292,28 @@ export default class ButtonView extends View<HTMLButtonElement> implements Butto
...
 		};
 
+		this.bind( '_ariaPressed' ).to(
+			this, 'isOn', this, 'isToggleable', this, 'role',
+			( isOn, isToggleable, role ) => {
+				if ( !isToggleable || isCheckableRole( role ) ) {
+					return false;
+				}
+
+				return String( !!isOn );
```
[the remark this answers is not in the document]
Comment (Mati365) [reply packages/ckeditor5-ui/src/button/buttonview.ts:302]: Yes, if `false` is returned, the attribute is not set to element.

File: packages/ckeditor5-ui/src/button/listitembuttonview.ts [hunk 5]
```diff
@@ -0,0 +1,195 @@
...
+		this.on<ObservableChangeEvent<boolean>>(
+			'change:_shouldRenderCheckHolder',
+			( evt, propertyName, shouldRenderCheckHolder ) => {
+				const { children, _checkIconHolderView } = this;
+
+				if ( shouldRenderCheckHolder && !children.has( _checkIconHolderView ) ) {
+					children.add( _checkIconHolderView, 0 );
+				} else if ( !shouldRenderCheckHolder && children.has( _checkIconHolderView ) ) {
+					children.remove( _checkIconHolderView );
+				}
```
[the remark this answers is not in the document]
Comment (Mati365) [reply packages/ckeditor5-ui/src/button/listitembuttonview.ts:106]: > Do you have 100% CC? Doesn't it complain about else path?

Yes, and it's tested here:
https://github.com/ckeditor/ckeditor5/blob/ad98d1fcfdec8b53f4ca3c3cedfd4356b1a0b6c1/packages/ckeditor5-ui/tests/button/listitembuttonview.js#L87-L94

> I ask because since you listen to change on boolean you have only two options false -> true or true -> false. So, based on new value, previous value is always deterministic and same for children state, right?

You are correct, removed.

File: packages/ckeditor5-ui/src/button/listitembuttonview.ts [hunk 6]
```diff
@@ -0,0 +1,195 @@
...
+			}
+		);
+	}
+}
+
+export class CheckIconHolderView extends View {
```
[the remark this answers is not in the document]
Comment (Mati365) [reply packages/ckeditor5-ui/src/button/listitembuttonview.ts:112]: @scofalik I exported it only for testing purpose. Is there a way to test this class without exporting it somewhere?

File: packages/ckeditor5-ui/src/button/listitembuttonview.ts [hunk 7]
```diff
@@ -0,0 +1,195 @@
...
+	private _watchCheckIconMount() {
+		this.on<ObservableChangeEvent<boolean>>( 'change:isOn', ( evt, propertyName, isOn ) => {
+			const { children, _checkIconView } = this;
+
+			if ( isOn && !children.has( _checkIconView ) ) {
+				children.add( _checkIconView );
+			} else if ( !isOn && children.has( _checkIconView ) ) {
+				children.remove( _checkIconView );
+			}
+		} );
```
[the remark this answers is not in the document]
Comment (Mati365) [reply packages/ckeditor5-ui/src/button/listitembuttonview.ts:177]: @scofalik Two things:

1. While it's a simpler approach, I'm not sure if it's the fastest one. Although the check is not rendered (due to `display: none`) it's still attached to the DOM. In smaller lists it makes no difference, but in the bigger lists it may. I assume so because SVG is not a trivial component to render, it's quite complicated, and our UI is not so performant in terms of rendering views.
2. Not every `ListItemButtonView` is checkbox. I don't think we should render unnecessary DOM elements in elements that will never become checkboxes, even if it's simpler, because it can be misleading from an inspector perspective.

File: packages/ckeditor5-ui/src/colorselector/colorgridsfragmentview.ts [hunk 8]
```diff
@@ -259,16 +259,21 @@ export default class ColorGridsFragmentView extends View {
 		if ( this.documentColorsCount ) {
 			// Create a label for document colors.
 			const bind = Template.bind( this.documentColors, this.documentColors );
-			const label = new LabelView( this.locale );
-			label.text = this._documentColorsLabel;
-			label.extendTemplate( {
+			const label = new View( this.locale );
```
[the remark this answers is not in the document]
Comment (Mati365) [reply packages/ckeditor5-ui/src/colorselector/colorgridsfragmentview.ts:262]: @scofalik It fixes:

> Document colors region in the color panel is incorrectly marked up as label

It's reported in this issue: https://github.com/cksource/ckeditor5-commercial/issues/6037#issue-2164994134. It looks like I picked too many changes from that PR. Although it's partially out of scope of this PR, but change is so small that there is no difference if it's kept here or in another PR. Can we keep it?

File: packages/ckeditor5-ui/src/dropdown/utils.ts [hunk 9]
```diff
@@ -541,6 +543,24 @@ function bindViewCollectionItemsToDefinitions(
...
+		const listItemButtons = [ ...listItems ].reduce<Array<ListItemButtonView>>( ( acc, item ) => {
+			if ( item instanceof ListItemView && item.children.first instanceof ListItemButtonView ) {
+				acc.push( item.children.first );
+			}
+
+			return acc;
+		}, [] );
```
[the remark this answers is not in the document]
Comment (Mati365) [reply packages/ckeditor5-ui/src/dropdown/utils.ts:555]: @scofalik This reduce performs filter and map at once. It's not plain `filter`.

File: packages/ckeditor5-ui/src/dropdown/utils.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (Mati365) [reply packages/ckeditor5-ui/src/dropdown/utils.ts:548]: > For performance reasons, I think we could add this event after the list is populated and also of course evaluate it once for initial load. I am a bit afraid of the longer lists, I am afraid it will have square complexity and may impact performance when a few hundred entries are loaded.

The performance should not be an issue, at least not in this particular listener. I checked it before implementation and `change` is fired [once](https://github.com/ckeditor/ckeditor5/blob/4f4032a74a8fe1ca0e4d6ace27bd660b095c674b/packages/ckeditor5-utils/src/collection.ts#L228-L234), event after calling `addMany`. So it's perfectly fine to assume that for the majority part of UI it'll be called only once, and it'll be fired occasionally while extending an existing dropdown, which is not super common across our UI. On the other hand - if there is and dropdown, that uses `add()` instead of `addMany()`, it can spam a bit and increase complexity. I don't think it matters.

File: packages/ckeditor5-ui/src/dropdown/utils.ts [hunk 10]
```diff
@@ -555,12 +575,16 @@ function bindViewCollectionItemsToDefinitions(
 
 			return groupView;
 		} else if ( def.type === 'button' || def.type === 'switchbutton' ) {
+			const isToggleable = def.model.role === 'menuitemcheckbox' || def.model.role === 'menuitemradio';
```
[the remark this answers is not in the document]
Comment (Mati365) [reply packages/ckeditor5-ui/src/dropdown/utils.ts:578]: @scofalik If you ask about general `Button` behavior, then it's for backward compatibility. There is huge amount of places where there are buttons created with `isToggleable` attribute, but there is no `role` set. Setting `isToggleable` to be controlled only by `role` attribute affects behavior of such buttons and might be a big breaking change which will kill a large amount of integrations. 

If you ask about this particular line, then it's `isToggleable` alone is not enough to display these buttons as checkboxes from the ARIA point of the view. We must know if it's checkbox or radio, and it has to be specified explicitly. If we ignore all aria things, then yes - we can use it, but I wanted to force programmers adding similar buttons with set `role` in the future. Average programmer, after checking "why my dropdown button is not checkbox!", after short debug figure out that he most likely forgot the `role` which is important in terms of ARIA.

File: packages/ckeditor5-ui/src/menubar/utils.ts [hunk 11]
```diff
@@ -163,6 +163,32 @@ export const MenuBarBehaviors = {
...
+
+	/**
+	 * Tracks the keyboard focus interaction on the menu bar view. It is used to determine if the nested items
+	 * of the menu bar should render focus rings after first interaction with the keyboard.
+	 */
+	trackKeyboardFocusInteraction( menuBarView: MenuBarView ): void {
+		let isKeyPressed: boolean = false;
+
+		menuBarView.on<ObservableChangeEvent<boolean>>( 'change:isOpen', () => {
+			menuBarView.hadKeyboardFocusInteraction = false;
```
[the remark this answers is not in the document]
Comment (Mati365) [reply packages/ckeditor5-ui/src/menubar/utils.ts:176]: > I think this should be only on close? 🤔.

I don't think if it makes any difference in this place. `hadKeyboardFocusInteraction` is respected only in submenus to hide or show focus on hover. In menubar list, it's completely ignored.

File: packages/ckeditor5-ui/src/menubar/menubarmenulistview.ts (hunk 3 above)
Comment (Mati365) [reply packages/ckeditor5-ui/src/menubar/menubarmenulistview.ts:42]: @scofalik There are a few things why it's so complicated and why I decided to not unify this behavior. At this moment, we have to:

1. Iterate over all items of menu and try to pick all list items buttons. If it's:
1.1. `ListItemButtonView` - then pick it as it is, it's safe to use `instanceof` here, because `ListItemButtonView` does not use `MenuBarMenuListView`
1.1. `ListSeparatorView`  - do nothing, rejected in `reduce`.
1.1. `MenuBarMenuView` there is circular reference, and we have to define a `isNestedMenuLikeView` guard that determines if it's something that has similar structure to `MenuBarMenuView` and has `buttonView`
2. If some of the buttons are checkboxes then force allocate list items space in all buttons.

As you see, it raises two problems:

1. Circular references in this file. We cannot use `instanceof MenuBarMenuView` because `MenuBarMenuView` imports this file itself. 
2. We cannot assume that the first item of `MenuBarMenuListItemView` is always the button. This view contains only `children` which can hold anything, so it's why there is filtering made. 

I think that this method should not be unified because menu is much more complex. The pending dropdown implementation implements it slightly differently too.

File: packages/ckeditor5-ui/src/menubar/utils.ts (hunk 1 above)
Comment (Mati365) [reply packages/ckeditor5-ui/src/menubar/utils.ts:187]: @scofalik Originally I implemented it using `KeyboardFocusTracker` that was observing every keyboard interaction on the document and then there were similar comparison for `isKeyPressed`. It worked fine, and it fixed the issue you mentioned with `enter` and `alt+f9`, and that one with focus jumping from another element to menu using keyboard, although after discussion it was not a great idea to add another global utility, and it'll be better to use it locally so I removed it. May it be better to keep this utility as global handler?

File: packages/ckeditor5-ui/src/button/buttonview.ts [hunk 12]
```diff
@@ -277,6 +292,28 @@ export default class ButtonView extends View<HTMLButtonElement> implements Butto
 			}
 		};
 
+		this.bind( '_ariaPressed' ).to(
+			this, 'isOn', this, 'isToggleable', this, 'role',
+			( isOn, isToggleable, role ) => {
+				if ( !isToggleable || isCheckableRole( role ) ) {
```
[the remark this answers is not in the document]
Comment (Mati365) [reply packages/ckeditor5-ui/src/button/buttonview.ts:298]: More info here https://github.com/ckeditor/ckeditor5/pull/16573#discussion_r1652031490

**Comment by Mati365** [comment]
> Other than what I commented, please try to use properties and methods name as short as possible. Maybe you'll find some room for improvement. 

I tend to use the most descriptive names as I can, even if they are longer, so we can read them as a recipe step that shortly describes what the method does without looking at docs or types. Although it's my personal opinion, I'm perfectly fine with renaming. Can you point me which methods and variables should be renamed (and what is a preferable way to name them?). It'd helpful and save my time.

> I also dislike maybeXXX approach. I get your point, but that's not really necessary if we have typed code and if the method is very short or you immediately early return like this:

Sorry, I forgot about our discussions about these `maybe` prefixes, and I renamed these `maybeXXX` variables. I used to do them while using optional monodic result of functions to quickly determine without looking at the type if type might be nullable.

**Comment by Mati365** [comment]
@scofalik 

> Below are conclusions from manual testing I did (some relate to commercial features PR) 

Thx, will be fixed.

**Comment by Mati365** [comment]
🚧  Work in progress

1. [ ] Following menu items are not getting "on" state: Link, Block quote, Comments archive, Accessibility dialog, Find and replace. Additionally, AI Assistant gets incorrect styling (it is all grayed, instead of getting a checkbox). Finally, Comments archive doesn't get "on" state in toolbar either -- it's not directly related to this ticket but we could fix it as it's the same part of the code.
2.  [x] Switch toggle menu items have incorrect styling. When they are on, they are blue. See track changes toolbar dropdown.
3. [ ]  Focus do not follow mouse hovering on the top-level menu items ("Insert", "View", etc.) IF the keyboard interaction was when the menu was not open. To test this focus menu using Alt+F9, then click right arrow, then hover other items.
4. [ ]  In revision history, in sidebar, for a selected revision (blue), if open the revision menu (click tripe-dot) the items inside have white font color.
5. [ ]  Format -> Highlight -> options, when one of them is selected it "loses its colors".

File: packages/ckeditor5-ui/src/menubar/utils.ts (hunk 11 above)
[the remark this answers is not in the document]
Comment (Mati365) [reply packages/ckeditor5-ui/src/menubar/utils.ts]: You were right, there were edge cases with `click` + `focus` listeners were intercepting such value, and after debugging it was easier to reset it on close.

File: packages/ckeditor5-ui/src/button/listitembuttonview.ts (hunk 7 above)
[the remark this answers is not in the document]
Comment (scofalik) [reply packages/ckeditor5-ui/src/button/listitembuttonview.ts:177]: Sure, that was just my thought, I thought that maybe we do such things somewhere anyway, so I thought to ask about it. I am completely fine with leaving it as it is, or maybe @oleq knows some cleaner way to manage it.

File: packages/ckeditor5-ui/src/colorselector/colorgridsfragmentview.ts (hunk 8 above)
[the remark this answers is not in the document]
Comment (scofalik) [reply packages/ckeditor5-ui/src/colorselector/colorgridsfragmentview.ts:262]: We can keep it, I just wanted to make sure that it's that, because I remember we had some `<label>` related issue.

File: packages/ckeditor5-ui/src/dropdown/utils.ts (hunk 9 above)
[the remark this answers is not in the document]
Comment (scofalik) [reply packages/ckeditor5-ui/src/dropdown/utils.ts:555]: Ah, the mythical filter-map. I didn't notice. Can we add an inline comment before this, something like:

```
// Filter-map. Out of all items, leave only these that have buttons and return the buttons.
```

File: packages/ckeditor5-ui/src/dropdown/utils.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (scofalik) [reply packages/ckeditor5-ui/src/dropdown/utils.ts:548]: If it uses `addMany()` then it is fine. I was worried the items are usually (always) added one by one.

File: packages/ckeditor5-ui/src/dropdown/utils.ts (hunk 9 above)
[the remark this answers is not in the document]
Comment (Mati365) [reply packages/ckeditor5-ui/src/dropdown/utils.ts:556]: Added

File: packages/ckeditor5-ui/src/dropdown/utils.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (scofalik) [reply packages/ckeditor5-ui/src/dropdown/utils.ts:548]: Can we add an inline comment:

```
// Possible theoretical performance problem if many items are added one by one, as this will be called for each item.
```

?

**Comment by Comandeer** [comment]
Replying to https://github.com/ckeditor/ckeditor5/pull/16573#issuecomment-2186700868:

"Focus visible" SC requires focus only during keyboard operations. I would argue that it allows us to skip showing the focus indicator if the user hovers over a top-level menu item with a mouse. Skipping the indicator there would make the whole experience consistent.

So:

1. If the user operates a top-level menu item with a mouse, do not add focus indicator.
2. As soon as the user starts using keyboard, show the focus indicator.

File: packages/ckeditor5-ui/src/dropdown/utils.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (Mati365) [reply packages/ckeditor5-ui/src/dropdown/utils.ts:550]: Comment added

File: packages/ckeditor5-ui/src/dropdown/utils.ts (hunk 10 above)
[the remark this answers is not in the document]
Comment (scofalik) [reply packages/ckeditor5-ui/src/dropdown/utils.ts:581]: I was asking in general. It seemed to me that if something is toggleable it should have one of the related ARIA roles. If we have big backward compatibility problem, then maybe we are missing a11y improvement in some places? :thinking:.

Anyway, we can leave this as a potential follow-up, if it is big as you say. You can resolve this conversation.

[the remark this answers is not in the document]
Comment (Mati365) [reply packages/ckeditor5-ui/src/dropdown/utils.ts:581]: @scofalik yep

File: packages/ckeditor5-ui/src/menubar/utils.ts (hunk 1 above)
Comment (scofalik) [reply packages/ckeditor5-ui/src/menubar/utils.ts:199]: I remember the original, implementation, hence I wrote:

> If we don't find a better and still reasonably simple solution then I am fine with that.

:slightly\_smiling\_face: 

In case of this particular topic (showing focus outline), keyboard interaction is a keyboard interaction that would lead to changing the menu state, including both focus and opening menus (and in that one case with enter, enter key does not change focus but I think "it counts").

But now as I think about it, the point where we should start tracking focus is actually Alt+F9 press. Because that's the first keyboard interaction that changes focus/menu state.

Isn't the solution as simple as just listening one element "higher", instead of on menu views, then on the menu bar?

**Comment by scofalik** [comment]
> Sorry, I forgot about our discussions about these `maybe` prefixes, and I renamed these `maybeXXX` variables. I used to do them while using optional monodic result of functions to quickly determine without looking at the type if type might be nullable.

:+1: 

> I tend to use the most descriptive names as I can, even if they are longer, so we can read them as a recipe step that shortly describes what the method does without looking at docs or types.

Of course, this is a good rule to follow. But sometimes these names get a bit too long and reading them line after line is tiring. Sometimes they do not fit convention. Some things I'd think about from this PR are:

*   Does it have to be Check**Holder** or maybe Check is enough?
*   `_shouldRenderCheckHolder` -> `_hasCheck`
*   `hasReservedCheckHolderSpace` -> `hasCheckSpace` or `requiresCheckSpace` or `addCheckSpace`

Do you have to change these names? Honestly, these ones are deep and used slightly, so it isn't a critical thing. But I think you have slight tendency to overdescribe the names, so you can treat it as a general advice.

File: packages/ckeditor5-ui/src/menubar/utils.ts (hunk 1 above)
Comment (Mati365) [reply packages/ckeditor5-ui/src/menubar/utils.ts:199]: > But now as I think about it, the point where we should start tracking focus is actually Alt+F9 press. 

@scofalik Yep, I discovered that in later commits and it looks like it works. Regarding `enter` issue - it was issue related to incorrect reset of `isKeyPressed` flag because some of the menus tends to intercept `keyup` and it was never freed.

**Comment by Mati365** [comment]
@scofalik Replaced 👍🏻

**Comment by scofalik** [comment]
One more example: `_preallocateCheckSpacerInItemsIfNeeded`. I am not a big fan of this name too :smile:. First, of course we do this only if needed, we don't do stuff if it is not needed. If you feel it is important, we can add proper comment in API doc. I'd probably go with something like `_setItemsCheckSpace`.

File: packages/ckeditor5-ui/src/menubar/menubarmenulistview.ts (hunk 3 above)
Comment (scofalik) [reply packages/ckeditor5-ui/src/menubar/menubarmenulistview.ts]: I wanted to analyze this and propose something but honestly, there's so much complication (mess) in the menu bar that I gave up. I think this feature far too many classes which extends too broad classes that allow for "any children" and the extended classes do not narrow down the children list.

Comment (Mati365) [reply packages/ckeditor5-ui/src/menubar/menubarmenulistview.ts]: @scofalik That's why dropdown menus operate mainly using trees

File: packages/ckeditor5-ui/src/menubar/utils.ts [hunk 13]
```diff
@@ -169,13 +169,25 @@ export const MenuBarBehaviors = {
...
+				isKeyPressed = false;
+			}
 		} );
 
+		menuBarView.listenTo( menuBarView.element!, 'click', () => {
+			if ( menuBarView.isOpen && menuBarView.element!.matches( ':focus-within' ) ) {
+				menuBarView.isFocusBorderEnabled = true;
+			}
+		}, { useCapture: true } );
+
```
Comment (scofalik) [inline packages/ckeditor5-ui/src/menubar/utils.ts:190]: This is incorrect. It reacts to every mouse click. E.g. I can click "Format" and then click "Text >" and it starts showing focus. On the other hand, it fixes the main issue a bit coincidentally. We shouldn't need to listen to a mouse click it all. After all, this is about keyboard interactions.

Instead Alt+F9 should cause `isFocusBorderEnabled = true` and all further iteractions should simply work. I guess that the problem is that when Alt+F9 happens, the focus is not in the menu, so listener below (`keydown`) does not trigger.

Let's summarize what works correctly now, and what needs to be revisited:

1.  When I mouse click a menu, it doesn't get a blue border anymore: correct.
2.  When I used keyboard to navigate menu (e.g. arrow down, enter), menu gets a blue border and it follows mouse hovers everywhere, including top-level menu: correct.
3.  When I Alt+F9 and click on the same top-level menu ("File"), it gets a blue border and border and focus follows mouse hover: correct but with incorrect fix. Also if I click another menu (e.g. "View") it works correctly.
4.  But when I Alt+F9 and start hovering other top-level menus, without clicking on them, the border does not move. What's more interesting, the focus does not move either! E.g. if I Alt+F9 and then hover a different item (e.g. "Insert") and press arrow right, the focus goes to "Edit" instead of "Format". This is different than e.g. when I click on the menu item (e.g. "Insert"), and then hover, then focus changes, and arrow keys work correctly. I don't know internals, so I don't know why there's a difference between Alt+F9 and click.

So, to summarize again:

1.  Alt+F9 should trigger `isFocusBorderEnabled`.
2.  Alt+F9 should causes focus to follow mouse.

File: packages/ckeditor5-basic-styles/src/utils.ts [hunk 14]
```diff
@@ -36,6 +36,13 @@ export function getButtonCreator( {
 		} );
 
 		view.bind( 'isEnabled' ).to( command, 'isEnabled' );
+		view.bind( 'isOn' ).to( command, 'value' );
+
+		if ( view instanceof ListItemButtonView ) {
```
[the remark this answers is not in the document]
Comment (scofalik) [reply packages/ckeditor5-basic-styles/src/utils.ts:41]: Also, it seems that we could move setting tooltip here as well.

**Review by scofalik — CHANGES_REQUESTED** [review]
Almost there. One bug left.

File: packages/ckeditor5-ui/src/menubar/utils.ts (hunk 13 above)
Comment (Mati365) [reply packages/ckeditor5-ui/src/menubar/utils.ts:190]: @scofalik 

> This is incorrect. It reacts to every mouse click. E.g. I can click "Format" and then click "Text >" and it starts showing focus. On the other hand, it fixes the main issue a bit coincidentally. We shouldn't need to listen to a mouse click it all. After all, this is about keyboard interactions.

It fixes edge case. If you open submenu using click it focuses first item of newly opened menu. It handles situation when user click button down while not having any item highlighted in menu.

Comment (Mati365) [reply packages/ckeditor5-ui/src/menubar/utils.ts:190]: @scofalik Alt+f9 is already handled and it should behave as you describe. I'll check that again but you have to have editor focused to make any keystrokes work

File: packages/ckeditor5-basic-styles/src/utils.ts (hunk 14 above)
[the remark this answers is not in the document]
Comment (Mati365) [reply packages/ckeditor5-basic-styles/src/utils.ts]: Changed

File: packages/ckeditor5-ui/src/menubar/utils.ts (hunk 13 above)
Comment (Mati365) [reply packages/ckeditor5-ui/src/menubar/utils.ts:190]: You are correct, I checked it on Windows VM and something is broken. I'm checking.

Comment (Mati365) [reply packages/ckeditor5-ui/src/menubar/utils.ts:197]: @scofalik Fixed, please reopen if it still occurs.

**Comment by oleq** [comment]
> * Since you can now close a tooltip using ESC, you need double ESC to close Format -> Bulleted List -> dropdown. If you wait a while, the first ESC will close the tooltip (e.g. "Circle" tooltip), then only the second ESC will close the dropdown. I found it counterintuitive and unexpected (cc @oleq).

Since this is the only place that acts this way, I wouldn't be worried. There are pros and cons of this approach.

> 5\. Icons in lists dropdowns do not get focus when hovered with mouse (cc @oleq).

AFAIR they are custom-made and don't share the same logic that makes up the menu bar. I don't know how to address this issue yet, though. Also, I suppose the same goes with in-toolbar dropdowns in general because they are different UI components and they are missing this mouse-to-keyboard logic.

**Comment by Mati365** [comment]
> AFAIR they are custom-made and don't share the same logic that makes up the menu bar. I don't know how to address this issue yet, though. Also, I suppose the same goes with in-toolbar dropdowns in general because they are different UI components and they are missing this mouse-to-keyboard logic.

I created separated issue about that: https://github.com/ckeditor/ckeditor5/issues/16612

File: packages/ckeditor5-ui/src/menubar/utils.ts [hunk 15]
```diff
@@ -182,6 +186,9 @@ export const MenuBarBehaviors = {
 			}
 		} );
 
+		// After clicking menu bar list item the focus is moved to the newly opened submenu.
+		// We need to enable focus border for the submenu items because after pressing arrow down it will
+		// focus second item instead of first which is not super intuitive.
```
Comment (scofalik) [inline packages/ckeditor5-ui/src/menubar/utils.ts:191]: After adding this comment I see your point now, and I think I agree with it. The crucial thing here is that we change the focus but this is not visualized and the blue border is the best way to visualize it.

Comment (Mati365) [reply packages/ckeditor5-ui/src/menubar/utils.ts:191]: Yep, exactly.

File: packages/ckeditor5-ui/src/menubar/utils.ts [hunk 16]
```diff
@@ -56,24 +56,28 @@ export const MenuBarBehaviors = {
 	 */
 	toggleMenusAndFocusItemsOnHover( menuBarView: MenuBarView ): void {
 		menuBarView.on<MenuBarMenuMouseEnterEvent>( 'menu:mouseenter', evt => {
-			// This works only when the menu bar has already been open and the user hover over the menu bar.
-			if ( !menuBarView.isOpen ) {
+			// This behavior should be activated when one of condition is present:
+			// 1. The user opened any submenu of menubar and hover over items in the menu bar.
+			// 2. The user focused whole menubar using keyboard interaction and enabled focus borders and hover over items in the menu bar.
+			if ( !menuBarView.isFocusBorderEnabled && !menuBarView.isOpen ) {
```
Comment (scofalik) [inline packages/ckeditor5-ui/src/menubar/utils.ts:62]: I don't see how these changes address the bug I was writing about so maybe the video will help.


https://github.com/ckeditor/ckeditor5/assets/1502228/ad477b03-5af0-47eb-bb12-5f12ffdfc734

At the beginning, I focus the editable, then I click Alt+F9. The menu bar gets focus. But after I move mouse around the menu bar, the focus is not moved. The blue outline does not follow the mouse and then I also click arrow keys, and they 1) don't start from where I have a mouse; 2) they don't trigger outline following the mouse.

Comment (scofalik) [reply packages/ckeditor5-ui/src/menubar/utils.ts:62]: I am sorry - this was my mistake - I pulled wrong repo. It works fine, thanks.

File: packages/ckeditor5-ui/src/button/listitembuttonview.ts (hunk 7 above)
[the remark this answers is not in the document]
Comment (scofalik) [reply packages/ckeditor5-ui/src/button/listitembuttonview.ts:177]: Okay, let's leave it as it is, with maybe one nitpick -- I think it is the same situation as with the other place, where `children.has` should be determined by previous `isOn` state. So maybe the `if` can be simplified.

[the remark this answers is not in the document]
Comment (Mati365) [reply packages/ckeditor5-ui/src/button/listitembuttonview.ts:177]: Ok, so I'm keeping it as it is

**Comment by scofalik** [comment]
Regarding tooltips: sometimes it is correct, sometimes not. I understand why we introduced closing tooltips on ESC and I don't discuss it - I agree that was an improvement. But in this case it was a UX degradation from my point of view and how I used it, I noticed it and wrote about it. There's probably no way to have both, so I am fine with how it is.

It wouldn't be a problem if the list components were better integrated with the menu bar, i.e., if I was able to leave the list types grid by pressing arrow left or arrow up when in left column / top row.
